src/state.rs

```diff
@@ -0,0 +1,214 @@
+use std::fmt;
+use std::marker::PhantomData;
+
+use primitives::{Error, RangeStream, Stream, StreamOnce};
+
+/// Trait for tracking the current position of a `Stream`.
+pub trait Positioner: PartialEq {
+    type Item;
+    /// The type which keeps track of the position
+    type Position: Clone + Ord;
+    /// Returns the current position
+    fn position(&self) -> Self::Position;
+    /// Updates the position given that `item` has been taken from the stream
+    fn update(&mut self, item: &Self::Item);
+}
+
+/// Trait for tracking the current position of a `RangeStream`.
+pub trait RangePositioner: Positioner {
+    type Range;
+    /// Updates the position given that `range` has been taken from the stream
+    fn update_range(&mut self, range: &Self::Range);
+}
+
+/// The `State<I>` struct maintains the current position in the stream `I` using
+/// the `Positioner` trait to track the position.
+///
+/// ```
+/// # extern crate combine;
+/// # use combine::{token, Parser, ParseError};
+/// # use combine::primitives::{Error};
+/// # use combine::state::{State, IndexPositioner};
+/// # fn main() {
+///     let result = token('9')
+///         .message("Not a nine")
+///         .parse(State::new("8", IndexPositioner::<char, &str>::new()));
+///     assert_eq!(result, Err(ParseError {
+///         position: 0,
+///         errors: vec![
+///             Error::Unexpected('8'.into()),
+///             Error::Expected('9'.into()),
+///             Error::Message("Not a nine".into())
+///         ]
+///     }));
+/// # }
+/// ```
+#[derive(Clone, PartialEq)]
+pub struct State<I, X>
+    where I: Stream,
+          X: Positioner<Position = I::Position>,
```

Comment:
While this was in my example, I don't think this restriction is actually necessary. If `I == &[T]` it would be nice to use `State` + `Positioner` to use something else for a position than `usize`.